include/rcutils/directory_and_file_reader.h

```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCUTILS__DIRECTORY_AND_FILE_READER_H_
+#define RCUTILS__DIRECTORY_AND_FILE_READER_H_
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcutils/allocator.h"
+#include "rcutils/types/rcutils_ret.h"
+#include "rcutils/macros.h"
+#include "rcutils/visibility_control.h"
+
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#ifndef _WIN32
+#include <dirent.h>
+#include <limits.h>
+# define RCUTILS_DIR_PATH_MAX PATH_MAX
+#else
+#include <windows.h>
+#include <tchar.h>
+#include <direct.h>
+# define RCUTILS_DIR_PATH_MAX MAX_PATH
+#endif  // _WIN32
+
+#ifndef RCUTILS_DIR_PATH_MAX
+#define RCUTILS_DIR_PATH_MAX 4096
+#endif
+#define RCUTILS_FILENAME_MAX 256
+/// Handle directories library.
+
+/// Handle files.
+typedef struct rcutils_file_t
+{
+  /// path of the file
+  char * path;
+  /// filename
+  char * name;
+  /// is the file a directory ?
+  int is_dir;
+  //  information about the file
+  #ifndef _WIN32
+  struct stat _s;
+  #else
+  struct _stat _s;
+  #endif
+} rcutils_file_t;
+
+/// Handle directories.
+typedef struct rcutils_dir_t
+{
+  /// path of the directory
+  char * path;
+  // is there any file in the folder.
+  int has_next;
+
+  #ifndef _WIN32
+  /// Directory struct
+  DIR * dir;
+  struct dirent * f;
+  #else
+  HANDLE dir;
+  WIN32_FIND_DATA f;
+  #endif
+
+  /// allocator
+  rcutils_allocator_t allocator;
+} rcutils_dir_t;
+
+
+/// Return an empty dir struct.
+/*
+ * This function returns an empty and zero initialized directory struct.
+ *
+ * Example:
+ * // Do not do this:
+ * // rcutils_dir_t foo;
+ * // rcutils_open_dir(&foo, "directory", rcutils_get_default_allocator()); // undefined behavior!
+ * // rcutils_ret_t ret = retrcutils_next_dir(&foo); // undefined behavior!
+ *
+ * // Do this instead:
```

Comment:
I would lead with the correct example, then provide negative examples.  I'm guilty of copying the first bit of code I see in the documentation.

Reply:
changed the order of the examples https://github.com/ros2/rcutils/pull/218/commits/1ff3651bb7398794bee1da36510f5d1abe28bbfa